### PR TITLE
Add GCN tutorial

### PR DESCRIPTION
**Pull request by Derek-Wds**
This PR is related to [#80](https://github.com/sql-machine-learning/models/pull/80). If there is any problem, please let me know. 

Thanks.

### CODE REVIEW

File: doc/tutorial/cora-gcn.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
+# Classify Cora Dataset Using GCN
+
+This tutorial shows how to train [GCN](https://arxiv.org/pdf/1609.02907.pdf) model on the Cora dataset. In this tutorial, you will learn how to:
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: train a GCN model ...

File: doc/tutorial/cora-gcn.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+- Train a GCN on the [Cora citation dataset](https://linqs-data.soe.ucsc.edu/public/lbc/cora.tgz).
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: Train a GCN model ...

File: doc/tutorial/cora-gcn.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+- Use the trained GCN to predict the label for some of the papers in the dataset.
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: Use the trained GCN model

File: doc/tutorial/cora-gcn.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+
+## The Dataset
+
+The Cora dataset is graph dataset about a citation network of scientific papers. It consists of 2708 scientific publications classified into one of seven classes. The citation network consists of 5429 links. Each publication in the dataset is described by a 0/1-valued word vector indicating the absence/presence of the corresponding word from the dictionary. The dictionary consists of 1433 unique words.
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: The Cora dataset is a graph dataset

File: doc/tutorial/cora-gcn.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+Due to the fact that graph data is not applicable to be divided into batches, we are not able to split the data and store in training dataset and test dataset respectively. Thus, we represent the entire graph through two tables: `Node Table` and `Edge Table`. (Anyone who wants to store graph data in the database can refer to this method.)
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: store in the training dataset ...

File: doc/tutorial/cora-gcn.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+-- | -- 
+id| Id for the node.
+name| Name for the node.
+features| Feature vector of the node represented in the string. An example would be "0,0,1".
+label | Label for the node.
+
+The following command can be used to construct the `Node Table`. 
+
+```sql
+CREATE TABLE cora.node (
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md:26]: All `sql` code blocks under `doc/tutorial` will be converted to a Jupyter Notebook in the released SQLFlow docker images, so we may need to add `%%sqlflow` in all `sql` blocks, and, if you want readers do not execute this block (the dataset is already loaded in the Docker image), try not to use `sql` block.

File: doc/tutorial/cora-gcn.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+In order to do so, you need to use the `JOIN` command in SQL to select all the data from `Node Table` and `Edge Table`. The following command is used to load all the data for training GCN.
+
+```sql
+%%sqlflow
+SELECT cora.node.id, features, label as class, cora.edge.from_node_id, cora.edge.to_node_id FROM cora.node
+LEFT JOIN cora.edge ON (cora.node.id = cora.edge.from_node_id OR cora.node.id = cora.edge.to_node_id)
+ORDER BY cora.node.id
+```
+The `OR` statement in the command is used to select all the possible bidirectional edges from the dataset. Without this `OR` command, some of the edges will be missing and it is not possible to construct the entire graph.
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: Add an empty line after the code block.

File: doc/tutorial/cora-gcn.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+### Load data from Cora
+You can load the data from database following a standard SQL command such as `SELECT * FROM cora.node`. However, since GCN is supposed to deal with graph data, you have to load both the `Node Table` and `Edge Table` at once.
+
+In order to do so, you need to use the `JOIN` command in SQL to select all the data from `Node Table` and `Edge Table`. The following command is used to load all the data for training GCN.
+
+```sql
+%%sqlflow
+SELECT cora.node.id, features, label as class, cora.edge.from_node_id, cora.edge.to_node_id FROM cora.node
+LEFT JOIN cora.edge ON (cora.node.id = cora.edge.from_node_id OR cora.node.id = cora.edge.to_node_id)
+ORDER BY cora.node.id
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: Append `;` at the end of the statement.

File: doc/tutorial/cora-gcn.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+```sql
+%%sqlflow
+TO TRAIN sqlflow_models.GCN
+WITH model.nhid=16, 
+     model.nclass=7, 
+     model.epochs=200, 
+     model.train_ratio=0.15, 
+     model.eval_ratio=0.2, 
+     validation.metrics="CategoricalAccuracy"
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md:116]: This statement can not be executed directly if this document is converted to a Jupyter Notebook, try put it together with the `SELECT` statement.

File: doc/tutorial/cora-gcn.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+You can specify the model parameters and training configurations through the `WITH` clause. For instance, you could set the `epochs` to be trained to be 100. `train_ratio` and `eval_ratio` indicate the proportion of the dataset to used for training and evaluate respectively. You can also change configurations such as `nlayer` which decides the number `GCNLayer` to be used, `dropout` which defines the dropout rate of the model.
```
Comment (typhoonzero) [inline doc/tutorial/cora-gcn.md]: Do you mean `model.epochs`?

File: doc/tutorial/cora-gcn.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+The Cora dataset is graph dataset about a citation network of scientific papers. It consists of 2708 scientific publications classified into one of seven classes. The citation network consists of 5429 links. Each publication in the dataset is described by a 0/1-valued word vector indicating the absence/presence of the corresponding word from the dictionary. The dictionary consists of 1433 unique words.
+
+Due to the fact that graph data is not applicable to be divided into batches, we are not able to split the data and store in training dataset and test dataset respectively. Thus, we represent the entire graph through two tables: `Node Table` and `Edge Table`. (Anyone who wants to store graph data in the database can refer to this method.)
+
+Here are the column description of the `Node Table`:
+
+Column | Explain 
+-- | -- 
+id| Id for the node.
```
Comment (lhw362950217) [inline doc/tutorial/cora-gcn.md]: We'd better claim the field type.

File: doc/tutorial/cora-gcn.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+name| Name for the node.
+features| Feature vector of the node represented in the string. An example would be "0,0,1".
```
Comment (lhw362950217) [inline doc/tutorial/cora-gcn.md]: How about point out we are using `csv` format.

File: doc/tutorial/cora-gcn.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+ORDER BY cora.node.id
+```
+The `OR` statement in the command is used to select all the possible bidirectional edges from the dataset. Without this `OR` command, some of the edges will be missing and it is not possible to construct the entire graph.
+
+With the `COLUMN` clause provided, SQLFlow can handle the comma separated string `features` with command `COLUMN DENSE(features)`.
+
+The GCN in SQLFlow is able to build the entire graph automatically with input in the folloing order: `node.id`, `node.features`, `node.label`, `edge.from_node_id`, `edge.to_node_id`. Please make sure the order is correct in order to run GCN successfully.
+
+### Train GCN
+After loading the dataset, you would be able to train GCN with following command:
```
Comment (lhw362950217) [inline doc/tutorial/cora-gcn.md]: How about list all supported parameters in a table, and given a description each.

File: doc/tutorial/cora-gcn.md (hunk 10 above)
Comment (Derek-Wds) [reply doc/tutorial/cora-gcn.md]: Yes, I will try to list all the supported parameters of GCN in a table and explain each a bit.

File: doc/tutorial/cora-gcn.md (hunk 11 above)
Comment (Derek-Wds) [reply doc/tutorial/cora-gcn.md]: Sure, I will add it for both tables

File: doc/tutorial/cora-gcn.md (hunk 13 above)
Comment (Derek-Wds) [reply doc/tutorial/cora-gcn.md]: Sure, I'm working on it.

File: doc/tutorial/cora-gcn.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        from_node_id INT,
+        to_node_id  INT,
+        weight FLOAT);
+```
+
+You can have a quick peek of the data by running the following standard SQL statements.
+
+```sql
+%%sqlflow
+describe cora.node;
```
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: For the convention of the SQL statement, try to write `DESCRIBE core.node;`?

File: doc/tutorial/cora-gcn.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+describe cora.edge;
+```
+
+```sql
+%%sqlflow
+select * from cora.node limit 10;
```
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: Try to write `SELECT * FROM cora.node LIMIT 10;`?

File: doc/tutorial/cora-gcn.md [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+## Train GCN on the Cora dataset
```
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: Train the GCN Model on the Cora Dataset

File: doc/tutorial/cora-gcn.md [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+Let's train a GCN model!
+
+### Load data from Cora
+You can load the data from database following a standard SQL command such as `SELECT * FROM cora.node`. However, since GCN is supposed to deal with graph data, you have to load both the `Node Table` and `Edge Table` at once.
```
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: You can load the data from **the** database following a standard SQL command such as `SELECT * FROM cora.node`. However, since **the** GCN **model** is supposed to deal with **the** graph data, you have to load both the `Node Table` and `Edge Table` at once.

File: doc/tutorial/cora-gcn.md (hunk 7 above)
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: Without this `OR` statement...

File: doc/tutorial/cora-gcn.md [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+%%sqlflow
+SELECT cora.node.id, features, label as class, cora.edge.from_node_id, cora.edge.to_node_id FROM cora.node
+LEFT JOIN cora.edge ON (cora.node.id = cora.edge.from_node_id OR cora.node.id = cora.edge.to_node_id)
+ORDER BY cora.node.id
+```
+The `OR` statement in the command is used to select all the possible bidirectional edges from the dataset. Without this `OR` command, some of the edges will be missing and it is not possible to construct the entire graph.
+
+With the `COLUMN` clause provided, SQLFlow can handle the comma separated string `features` with command `COLUMN DENSE(features)`.
+
+The GCN in SQLFlow is able to build the entire graph automatically with input in the folloing order: `node.id`, `node.features`, `node.label`, `edge.from_node_id`, `edge.to_node_id`. Please make sure the order is correct in order to run GCN successfully.
```
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: The GCN model

with inputs

in order to run the GCN model

File: doc/tutorial/cora-gcn.md (hunk 13 above)
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: to train the GCN model

File: doc/tutorial/cora-gcn.md [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+After the prediction, you can check the prediction result by
+
+```sql
+%%sqlflow
+SELECT * FROM cora.predict LIMIT 5;
+```
+
+## Evaluate GCN
```
Comment (sneaxiy) [inline doc/tutorial/cora-gcn.md]: Evaluate the Trained GCN Model

File: doc/tutorial/cora-gcn.md (hunk 14 above)
Comment (Derek-Wds) [reply doc/tutorial/cora-gcn.md]: Sure, I will update it.

**Review by lhw362950217 — APPROVED** [review]
LGTM
